modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -57,67 +70,165 @@ public KafkaContainer(final DockerImageName dockerImageName) {
+                .computeIfAbsent(
+                    "KAFKA_CONTROLLER_QUORUM_VOTERS",
+                    key ->
+                        String.format(
+                            "%s@%s:9094",
+                            getEnvMap().get("KAFKA_NODE_ID"),
+                            getNetwork() != null ? getNetworkAliases().get(0) : "localhost"
+                        )
+                )
+        );
+        withEnv("KAFKA_CONTROLLER_LISTENER_NAMES", "CONTROLLER");
+    }
+
+    protected void configureZookeeper() {
         if (externalZookeeperConnect != null) {
             withEnv("KAFKA_ZOOKEEPER_CONNECT", externalZookeeperConnect);
         } else {
             addExposedPort(ZOOKEEPER_PORT);
             withEnv("KAFKA_ZOOKEEPER_CONNECT", "localhost:" + ZOOKEEPER_PORT);
-            command += "echo 'clientPort=" + ZOOKEEPER_PORT + "' > zookeeper.properties\n";
-            command += "echo 'dataDir=/var/lib/zookeeper/data' >> zookeeper.properties\n";
-            command += "echo 'dataLogDir=/var/lib/zookeeper/log' >> zookeeper.properties\n";
-            command += "zookeeper-server-start zookeeper.properties &\n";
         }
+    }
+
+    @Override
+    protected void containerIsStarting(InspectContainerResponse containerInfo) {
+        super.containerIsStarting(containerInfo);
+
+        String command = "#!/bin/bash\n";
+        // exporting KAFKA_ADVERTISED_LISTENERS with the container hostname
+        command +=
+            String.format(
+                "export KAFKA_ADVERTISED_LISTENERS=%s,%s\n",
+                getBootstrapServers(),
+                brokerAdvertisedListener(containerInfo)
+            );
+
+        command += (kraftEnabled) ? commandKraft() : commandZookeeper();
 
         // Optimization: skip the checks
         command += "echo '' > /etc/confluent/docker/ensure \n";
         // Run the original command
         command += "/etc/confluent/docker/run \n";
-        withCommand("sh", "-c", command);
+        copyFileToContainer(Transferable.of(command, 0777), STARTER_SCRIPT);
     }
 
-    @Override
-    @SneakyThrows
-    protected void containerIsStarted(InspectContainerResponse containerInfo) {
-        String brokerAdvertisedListener = brokerAdvertisedListener(containerInfo);
-        ExecResult result = execInContainer(
-            "kafka-configs",
-            "--alter",
-            "--bootstrap-server",
-            brokerAdvertisedListener,
-            "--entity-type",
-            "brokers",
-            "--entity-name",
-            getEnvMap().get("KAFKA_BROKER_ID"),
-            "--add-config",
-            "advertised.listeners=[" + String.join(",", getBootstrapServers(), brokerAdvertisedListener) + "]"
-        );
-        if (result.getExitCode() != 0) {
-            throw new IllegalStateException(result.toString());
-        }
+    protected String commandKraft() {
+        String command = "sed -i '/KAFKA_ZOOKEEPER_CONNECT/d' /etc/confluent/docker/configure\n";
```

Comment:
We can do something like this. That way, we don't need `generateClusterId()` and `KafkaContainerKraftCluster` can use `Uuid.randomUuid().toString()`. WDYT?

```suggestion
    protected String commandKraft() {
        try {
            if (clusterId == null) {
                clusterId = execInContainer("kafka-storage", "random-uuid").getStdout().trim();
            }
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
        String command = "sed -i '/KAFKA_ZOOKEEPER_CONNECT/d' /etc/confluent/docker/configure\n";
```

Reply:
Ok got it.
Indeed, it is very likely that `KafkaContainerKraftCluster` will have the `kafka-clients` dependency.